resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ResourceRequestHelper.scala

```diff
@@ -143,17 +143,28 @@ private object ResourceRequestHelper extends Logging {
     require(resource != null, "Resource parameter should not be null!")
 
     logDebug(s"Custom resources requested: $resources")
+    if (resources.isEmpty) {
+      // no point in going forward, as we don't have anything to set
+      return
+    }
+
     if (!isYarnResourceTypesAvailable()) {
-      if (resources.nonEmpty) {
-        logWarning("Ignoring custom resource requests because " +
-            "the version of YARN does not support it!")
-      }
+      logWarning("Ignoring custom resource requests because " +
+          "the version of YARN does not support it!")
       return
     }
 
     val resInfoClass = Utils.classForName(RESOURCE_INFO_CLASS)
     val setResourceInformationMethod =
-      resource.getClass.getMethod("setResourceInformation", classOf[String], resInfoClass)
+      try {
+        resource.getClass.getMethod("setResourceInformation", classOf[String], resInfoClass)
+      } catch {
+        case e: NoSuchMethodException =>
+          throw new SparkException(
+            s"Cannot find $RESOURCE_INFO_CLASS.setResourceInformation. " +
+              "This is likely due to a jar conflict between different yarn versions.", e)
```

Comment:
I actually think that the most common case will be people trying to use with the wrong version of the Hadoop libraries. The case you mention is only really a problem for Spark devs...

So the message should say the Hadoop library in use does not support resource requests, instead. (Spark devs should be smart enough to make the association... there's a bunch of these kinds of conflicts that cause all sorts of different exceptions.)

Reply:
Actually nevermind. It only happens if you mix jars. (ResourceInformation.java is pretty old but it's still hadoop-3 only.)

Anyway, you error message is wrong since the missing method is in the `Resource` class. Also, "YARN".

Reply:
good catch @vanzin, updated.